woocommerce/woocommerce-android
# [REST API] Handle WordPress.com username login and fallback when email is not supported

Closes: #10036  Closes: #10032


### Description
This PR adds two changes:
1. It allows using the username for signing in using WordPress.com.
2. It handles the case of `EMAIL_LOGIN_NOT_ALLOWED` by showing an error message that asks the user to use their username instead.

The above flow was discussed and approved here p1677851752776439/1677500033.040359-slack-C046HDZL87J

### Testing instructions
I couldn't find an account that's flagged for this test, and I wasn't able to create one, so let's just simulate this with code.

1. Apply the patch from below.
2. Open the app then sign in using Application Passwords to a non-Jetpack site.
3. Click on the Jetpack benefits banner in MyStore.
4. Attempt to sign in using an email.
5. Notice the error message: "Please log in using your WordPress.com username instead of your email address."
6. Try to sign in using your username.
7. Confirm the flow continues to the password correctly.
8. Enter a correct password and confirm the sign in works fine.

The Jetpack connection is currently broken, I created an https://github.com/woocommerce/woocommerce-android/issues/10043 about it, so the rest of the flow won't work for your probably.

```patch
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/WPComLoginRepository.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/WPComLoginRepository.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/WPComLoginRepository.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/WPComLoginRepository.kt	(revision 868f607db76cebcba8b839f69d95a4b3fc8a3206)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/WPComLoginRepository.kt	(date 1698417481826)
@@ -1,5 +1,6 @@
 package com.woocommerce.android.ui.login
 
+import androidx.core.util.PatternsCompat
 import com.woocommerce.android.OnChangedException
 import com.woocommerce.android.util.WooLog
 import com.woocommerce.android.util.dispatchAndAwait
@@ -9,6 +10,8 @@
 import org.wordpress.android.fluxc.network.rest.wpcom.auth.AccessToken
 import org.wordpress.android.fluxc.store.AccountStore.AuthEmailPayload
 import org.wordpress.android.fluxc.store.AccountStore.AuthEmailPayloadScheme
+import org.wordpress.android.fluxc.store.AccountStore.AuthOptionsError
+import org.wordpress.android.fluxc.store.AccountStore.AuthOptionsErrorType
 import org.wordpress.android.fluxc.store.AccountStore.AuthenticatePayload
 import org.wordpress.android.fluxc.store.AccountStore.AuthenticationError
 import org.wordpress.android.fluxc.store.AccountStore.AuthenticationErrorType.NEEDS_2FA
@@ -24,6 +27,10 @@
     private val wpComAccessToken: AccessToken
 ) {
     suspend fun fetchAuthOptions(emailOrUsername: String): Result<AuthOptions> {
+        if (PatternsCompat.EMAIL_ADDRESS.matcher(emailOrUsername).matches()) {
+            return Result.failure(OnChangedException(AuthOptionsError(AuthOptionsErrorType.EMAIL_LOGIN_NOT_ALLOWED, "")))
+        }
+
         val action = AccountActionBuilder.newFetchAuthOptionsAction(
             FetchAuthOptionsPayload(emailOrUsername)
         )
```
